File: Chapter-5/heap_sort.cpp
#include <iostream>
using namespace std;

void Heapify(int ar[], int num, int i){
    int largest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;

    if(l<num && ar[l] > ar[largest])
      largest = l;

    if(r< num && ar[r] > ar[largest])
      largest = r;

    if(largest != i) {
        swap(ar[i], ar[largest]);
        Heapify(ar, num, largest);
    }
}

void Heap_Sort(int ar[], int num){
    for(int i = num / 2 - 1; i >= 0; i--)
        Heapify(ar, num, i);

    for(int i = num - 1; i >= 0; i--){
        swap(ar[0], ar[i]);

        Heapify(ar, i, 0);
    }
}

void Display(int ar[], int num){
    cout<<endl<<"Sorted array using Heap Sort : ";
    for (int i = 0; i < num; ++i)
        cout << ar[i] << " ";
}

int main() {
    int num, i;
	cout<<" enter the number of elements in the array : ";
	cin>>num;

	int ar[num];
	cout<<"enter the elements: ";
	for(i = 0; i < num; i++){
		cin>>ar[i];
	}
    Heap_Sort(ar, num);
    Display(ar, num);
}

File: Chapter-5/heap_sort.py
def Heapify(a, num, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < num and a[i] < a[l]:
        largest = l

    if r < num and a[largest] < a[r]:
        largest = r

    if largest != i:
        a[i], a[largest] = a[largest], a[i]
        Heapify(a, num, largest)


def Heap_Sort(a):
    num = len(a)

    for i in range(num // 2, -1, -1):
        Heapify(a, num, i)

    for i in range(num - 1, 0, -1):
        a[i], a[0] = a[0], a[i]
        Heapify(a, i, 0)


a = []
num = int(input("enter the number of elements in the array : "))
print("Enter the elements : ")
for i in range(0, num):
    elements = int(input())
    a.append(elements)
Heap_Sort(a)
print(" Sorted Array using Heap Sort is : ", end = " ")
for i in range(num):
    print(a[i], end=" ")
